Default value are applied for empty string when there's no existing state.

### Terraform Version
```
$ terraform version
Terraform v0.12.2

$ grep terraform go.mod
module terraform-provider-fake
	github.com/hashicorp/terraform v0.12.2
```

Hi,

I'm a maintainer of a Terraform provider and I have a strange behavior with default value on string field (at least) when switching a provider to Terraform 1.12 (not sure if it's expected or not)

Default value (via `Default` or `DefaultFunc`) is applied when a string field is set as an empty string but only on the first plan/apply (when there's not existing state)

I created a [little fake provider](https://github.com/cyrilgdn/terraform-provider-fake) to help reproduce this behavior hopping it can help you.

It contains only the [following resource](https://github.com/cyrilgdn/terraform-provider-fake/blob/master/resource_fake.go):

```go
func resourceFake() *schema.Resource {
	return &schema.Resource{
		Create: createFake,
		Read:   readFake,
		Update: createFake,
		Delete: deleteFake,

		Schema: map[string]*schema.Schema{
			"value": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Default:  "plop",
			},
		},
	}
}

func createFake(d *schema.ResourceData, m interface{}) error {
	d.SetId("toto")
	return nil
}

func readFake(d *schema.ResourceData, m interface{}) error {
	return nil
}

func deleteFake(d *schema.ResourceData, m interface{}) error {
	return nil
}
```

## Steps to reproduce

 * Steps to reproduce are available in the [fake provider README](https://github.com/cyrilgdn/terraform-provider-fake) too
* There's also a full output for this example in a [Travis build](https://travis-ci.org/cyrilgdn/terraform-provider-fake/builds/546979940)

### Build the provider
```bash
git clone git@github.com:cyrilgdn/terraform-provider-fake.git
cd terraform-provider-fake
go build
```

Then with the following `main.tf` file:

```hcl
resource "fake" "test" {
  value = ""
}
```

Starting from scratch (with no state), **default value is applied**:

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be created
  + resource "fake" "test" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
    value = "plop"
}

```

Updating `main.tf` with: 
```hcl
resource "fake" "test" {
  value = "tmp"
}
```

It applies this new value: 

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be updated in-place
  ~ resource "fake" "test" {
        id    = "toto"
      ~ value = "plop" -> "tmp"
    }

Plan: 0 to add, 1 to change, 0 to destroy.


$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
    value = "tmp"
}
```

Reverting `main.tf` to set an empty string again:

```hcl
resource "fake" "test" {
  value = ""
}
```

Default value is **not applied** (as expected?) so field is cleared:

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be updated in-place
  ~ resource "fake" "test" {
        id    = "toto"
      - value = "tmp" -> null
    }

Plan: 0 to add, 1 to change, 0 to destroy.


$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
}
```


### With null value it seems to work as expected

```hcl
resource "fake" "test" {
  value = null
}
```

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be created
  + resource "fake" "test" {
      + id    = (known after apply)
      + value = "plop"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
    value = "plop"
}
```

```hcl
resource "fake" "test" {
  value = "tmp"
}
```

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be updated in-place
  ~ resource "fake" "test" {
        id    = "toto"
      ~ value = "plop" -> "tmp"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
    value = "tmp"
}
```

```hcl
resource "fake" "test" {
  value = null
}
```

```bash
$ terraform plan
Terraform will perform the following actions:

  # fake.test will be updated in-place
  ~ resource "fake" "test" {
        id    = "toto"
      ~ value = "tmp" -> "plop"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

$ terraform apply -auto-approve
[...]

$ terraform state show fake.test
# fake.test:
resource "fake" "test" {
    id    = "toto"
    value = "plop"
}
```


### In Terraform 0.11, default values are not applied for empty string even when there's no existing state.

```bash
$ rm .terraform terraform.tfstate -r
$ sed -i "s/0.12.2/0.11.7/" go.mod
$ go build
$ terraform011 init
```

With this `main.tf`
```hcl
resource "fake" "test" {
  value = ""
}
```

```bash
$ terraform011 plan 
Terraform will perform the following actions:

  + fake.test
      id: <computed>


Plan: 1 to add, 0 to change, 0 to destroy.


$ terraform011 apply -auto-approve
[...]

$ terraform011 state show fake.test
id    = toto
```

Updating and reverting:

```hcl
resource "fake" "test" {
  value = "tmp"
}
```

```bash

$ terraform plan
Terraform will perform the following actions:

  ~ fake.test
      value: "" => "tmp"


Plan: 0 to add, 1 to change, 0 to destroy.

$ terraform011 apply -auto-approve
[...]

$ terraform011 state show fake.test
id    = toto                
value = tmp
```

```hcl
resource "fake" "test" {
  value = ""
}
```

```bash

$ terraform plan
Terraform will perform the following actions:

  ~ fake.test
      value: "tmp" => ""

Plan: 0 to add, 1 to change, 0 to destroy.

$ terraform011 apply -auto-approve
[...]

$ terraform011 state show fake.test
id    = toto                          
value = 
```

I hope I'm not missing something and don't creating an issue for nothing.
Thanks in advance for your help in any case.

I've just created an [acceptance test](https://github.com/cyrilgdn/terraform-provider-fake/blob/master/resource_fake_test.go) on this fake provider which show this behavior.

```go
func TestAccFake__empty(t *testing.T) {
        config := `
resource "fake" "test" {
  value = ""
}
`
        resource.Test(t, resource.TestCase{
                Providers: testAccProviders,
                Steps: []resource.TestStep{
                        {
                                Config: config,
                                Check: resource.ComposeTestCheckFunc(
                                        resource.TestCheckResourceAttr("fake.test", "id", "toto"),
                                        resource.TestCheckResourceAttr("fake.test", "value", ""),
                                ),
                        },
                },
        })
}

func TestAccFake__default_value(t *testing.T) {
        config := `
resource "fake" "test" {
  value = ""
}
`
        resource.Test(t, resource.TestCase{
                Providers: testAccProviders,
                Steps: []resource.TestStep{
                        {
                                Config: config,
                                Check: resource.ComposeTestCheckFunc(
                                        resource.TestCheckResourceAttr("fake.test", "id", "toto"),
                                        resource.TestCheckResourceAttr("fake.test", "value", "plop"),
                                ),
                        },
                },
        })
}
```

Running the test:
```bash
$ make testacc
TF_ACC=true go test -v terraform-provider-fake
=== RUN   TestAccFake__empty
--- FAIL: TestAccFake__empty (0.02s)
    testing.go:568: Step 0 error: Check failed: Check 2/2 error: fake.test: Attribute 'value' expected "", got "plop"
=== RUN   TestAccFake__default_value
--- FAIL: TestAccFake__default_value (0.02s)
    testing.go:568: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: fake.test
          id:    "toto" => "toto"
          value: "plop" => ""
        
        
        
        STATE:
        
        fake.test:
          ID = toto
          provider = provider.fake
          value = plop
FAIL
FAIL	terraform-provider-fake	0.047s
make: *** [Makefile:2: testacc] Error 1
```

We can see that:
* If I expect to have an empty string, it's actually the default value.
* If I expect to have the default value, applying the step again will create a non-empty plan changing my value to an empty string.

--- a/builtin/providers/test/resource_defaults_test.go
+++ b/builtin/providers/test/resource_defaults_test.go
@@ -174,3 +174,22 @@ resource "test_resource_defaults" "foo" {
 		},
 	})
 }
+
+func TestDefaults_emptyString(t *testing.T) {
+	config := `
+resource "test_resource_defaults" "test" {
+  default_string = ""
+}
+`
+	resource.UnitTest(t, resource.TestCase{
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_defaults.test", "default_string", ""),
+				),
+			},
+		},
+	})
+}

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -1201,6 +1201,8 @@ func normalizeNullValues(dst, src cty.Value, apply bool) cty.Value {
 		}
 	}
 
+	// check the invariants that we need below, to ensure we are working with
+	// non-null and known values.
 	if src.IsNull() || !src.IsKnown() || !dst.IsKnown() {
 		return dst
 	}
@@ -1319,8 +1321,12 @@ func normalizeNullValues(dst, src cty.Value, apply bool) cty.Value {
 			return cty.ListVal(dsts)
 		}
 
-	case ty.IsPrimitiveType():
-		if dst.IsNull() && src.IsWhollyKnown() && apply {
+	case ty == cty.String:
+		// The legacy SDK should not be able to remove a value during plan or
+		// apply, however we are only going to overwrite this if the source was
+		// an empty string, since that is what is often equated with unset and
+		// lost in the diff process.
+		if dst.IsNull() && src.AsString() == "" {
 			return src
 		}
 	}

--- a/helper/plugin/grpc_provider_test.go
+++ b/helper/plugin/grpc_provider_test.go
@@ -985,14 +985,39 @@ func TestNormalizeNullValues(t *testing.T) {
 		{
 			// Plan primitives are kept
 			Src: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.NumberIntVal(0),
+			}),
+			Dst: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.NullVal(cty.Number),
+			}),
+			Expect: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.NullVal(cty.Number),
+			}),
+		},
+		{
+			// Neither plan nor apply should remove empty strings
+			Src: cty.ObjectVal(map[string]cty.Value{
 				"string": cty.StringVal(""),
 			}),
 			Dst: cty.ObjectVal(map[string]cty.Value{
 				"string": cty.NullVal(cty.String),
 			}),
 			Expect: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.StringVal(""),
+			}),
+		},
+		{
+			// Neither plan nor apply should remove empty strings
+			Src: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.StringVal(""),
+			}),
+			Dst: cty.ObjectVal(map[string]cty.Value{
 				"string": cty.NullVal(cty.String),
 			}),
+			Expect: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.StringVal(""),
+			}),
+			Apply: true,
 		},
 		{
 			// The null map is retained, because the src was unknown

